src/poll.c

```diff
@@ -0,0 +1,372 @@
+
+	poll_thread = pthread_self();
+
+	// Recieve signal USR2 and add a signal handler.
+	{
+		sigset_t sigset;
+		sigemptyset(&sigset);
+		sigaddset(&sigset, SIGUSR2);
+
+		if(pthread_sigmask(SIG_UNBLOCK, &sigset, NULL) == -1) {
+			error("Could not unblock USR2 for the polling thread");
+		}
+
+		struct sigaction sa;
+		sigemptyset(&sa.sa_mask);
+		sa.sa_handler = poll_handler;
+		sa.sa_flags = 0;
+		if(sigaction(SIGUSR2, &sa, NULL) == -1) {
+			error("Failed to change signal handler for SIGHUP");
+		}
+	}
+
+	fd_set readfds, writefds, exceptfds;
+	struct poll_file *p;
+	int nfds;
+
+	// Infinite Poll
+	while(1) {
+		FD_ZERO(&readfds);
+		FD_ZERO(&writefds);
+		FD_ZERO(&exceptfds);
+		nfds = 0;
+
+		//-------------------------------------------------------------------------
+		// Initialize select
+		poll_file_list_lock();
+		for(p = poll_file_head; p ; p = p->next) {
+			switch(p->type) {
+				case POLLIN:
+					FD_SET(p->fd, &readfds);
+					break;
+				case POLLOUT:
+					FD_SET(p->fd, &writefds);
+					break;
+				case POLLERR:
+					FD_SET(p->fd, &exceptfds);
+					break;
+				default: continue;
+			}
+			if(p->fd > nfds)
+				nfds = p->fd;
+		}
+		poll_file_list_unlock();
+		//-------------------------------------------------------------------------
+
+		select(nfds + 1, &readfds, &writefds, &exceptfds, NULL);
+
+		//-------------------------------------------------------------------------
+		// Handle polled files
+		poll_file_list_lock();
+		for(p = poll_file_head; p ; p = p->next) {
+			switch(p->type) {
+				case POLLIN:
+					if(FD_ISSET(p->fd, &readfds)) {
+						poll_time_update_nolock(&p->tv);
+					}
+					break;
+				case POLLOUT:
+					if(FD_ISSET(p->fd, &writefds)) {
+						poll_time_update_nolock(&p->tv);
+					}
+					break;
+				case POLLERR:
+					if(FD_ISSET(p->fd, &exceptfds)) {
+						poll_time_update_nolock(&p->tv);
+					}
+					break;
+			}
+		}
+		poll_file_list_unlock();
+		//-------------------------------------------------------------------------
+	}
+}
+
+// ---------------------------------------------------------------------------
+// API - basic methods
+void *poll_file_register(char *path, int type) {
+	struct poll_file *p;
+
+	// Search if we already poll this file.
+	poll_file_list_lock();
+	p = poll_file_list_search_nolock(path, type);
+	poll_file_list_unlock();
+
+	// If not found, init and add poll_file.
+	if(!p) {
+		p = poll_file_init(path, type);
+		poll_file_list_lock();
+		poll_file_list_prepend_nolock(p);
+		poll_file_list_unlock();
+	}
+
+	// Add a new checker
+	poll_file_list_lock();
+	struct poll_check *retv = poll_check_init_nolock(p);
+	poll_file_list_unlock();
+
+	// Notify poll thread to start watching the new file
+	poll_interrupt();
+
+	return retv;
+}
+
+unsigned long long poll_occured(void *poll_descriptor) {
+	struct poll_check *p_check = poll_descriptor;
+	unsigned long long retv;
+
+	poll_file_list_lock();
+	retv = poll_time_difference_nolock(&p_check->poll_file->tv, &p_check->tv);
```

Comment:
I think no locking is needed here. The whole idea of locking is to protect the linked list. We could assume that for each poll entry, the owner is only one thread. So, there is no way to delete an entry while the caller access it.

If there is the possibility of multiple threads accessing the same poll, even the `void *` we return might be invalid.